src/sync/shared_memory.zig

```diff
@@ -0,0 +1,148 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+
+const Allocator = std.mem.Allocator;
+
+/// Thread safe Window that stores a single copy of data that is shared with
+/// readers as a pointer to the underlying data inside the Window.
+///
+/// - this struct owns the data and is responsible for freeing it
+/// - the lifetime of returned pointer exceeds every read operation of that pointer,
+///   even if another thread evicts it from the Window, as long as `release` is used properly.
+pub fn SharedPointerWindow(
+    T: type,
+    deinitItem_: anytype,
+    DeinitContext: type,
+) type {
+    const Window = sig.utils.collections.Window;
+    const Rc = sig.sync.Rc;
+    const deinitItem = normalizeDeinitFunction(T, DeinitContext, deinitItem_);
+
+    return struct {
+        allocator: Allocator,
+        window: Window(Rc(T)),
+        center: std.atomic.Value(usize),
+        lock: std.Thread.RwLock = .{},
```

Comment:
Unless explicitly required I prefer the use of a RwLock wrapper around the necessary data which in this instance is `window` as far as I understand. I understand this is a matter of taste to some extent though so am ok with leaving it as is if you feel more strongly about it.

Reply:
I agree and I would argue that I am already using the pattern that you are advocating. SharedWindowPointer is the locking wrapper struct which wraps the Window type. The only reason this struct exists is to serve as an alternative to RwMux that provides stronger guarantees about synchronization correctness.

Reply:
The `lock` held by the struct is primarily used to control synchronisation on the `window` field. Changing the `window` field type to `RwMux(Window(Rc(T)))` reflects that fact in the type system. I have noticed a preference for this approach in previous PR reviews as well so thought it worth a mention.  At the end of the day I don't feel strongly enough to hold things up so feel free to resolve this comment.

Reply:
You could make the same argument about the inner fields within RwMux. RwMux contains a RwLock to protect a specific field that it contains. Obviously you wouldn't wrap the field inside a RwMux with another struct that's just like RwMux, since that would be redundant.

This data structure is analogous to RwMux. I feel that the same reasoning applies.